### PR TITLE
[T7150]connector_woocommerce:Downloadable Product-BAD-PAA

### PR DESCRIPTION
**Pull request by bizzappdevteam**
Task description:-

- The downloadable product should be a "Service" as the product type.
- Take the "Downloadable Product" boolean of woo-commerce at the binding level.
- Create a new object "woo.downloadable.product" and put it in the submenu of Woocommerce.
- Take One2many of "woo.downloadable.product"

### CODE REVIEW

File: bad_connector_woocommerce/models/product/common.py [hunk 1]
```diff
@@ -126,6 +126,12 @@ class WooProductProduct(models.Model):
         string="Computed Quantity",
         help="""Last computed quantity to send " "on WooCommerce.""",
     )
+    downloadable_product = fields.Boolean()
```
Comment (bizzappdevteam) [inline bad_connector_woocommerce/models/product/common.py]: make it readonly

File: bad_connector_woocommerce/views/product_views.xml [hunk 2]
```diff
@@ -58,6 +58,7 @@
                                     <field name="price" />
                                     <field name="regular_price" />
                                     <field name="stock_management" />
+                                    <field name="downloadable_product" readonly="1" />
```
Comment (bizzappdevteam) [inline bad_connector_woocommerce/views/product_views.xml]: remove from here and add at py level and write test case for downloadable product.

File: bad_connector_woocommerce/models/woo_downloadable_product/common.py [hunk 3]
```diff
@@ -0,0 +1,10 @@
+from odoo import fields, models
+
+
+class WooProductImageUrl(models.Model):
+    _name = "woo.downloadable.product"
+    _description = "WooCommerce Downloadable Product"
+
+    name = fields.Char(string="File Name", required=True)
+    url = fields.Char(string="File URL", required=True)
```
Comment (bizzappdevteam) [inline bad_connector_woocommerce/models/woo_downloadable_product/common.py]: keep only name as required field

File: bad_connector_woocommerce/models/product/importer.py [hunk 4]
```diff
@@ -215,8 +221,74 @@ def _after_import(self, binding, **kwargs):
...
-            return result
-        image_importer = self.component(usage="product.image.importer")
-        image_importer.run(self.external_id, binding, image_record)
+        if image_record:
+            image_importer = self.component(usage="product.image.importer")
+            image_importer.run(self.external_id, binding, image_record)
+
+        download_product_record = self.remote_record.get("downloads", [])
+        if download_product_record:
+            download_product_importer = self.component(
```
Comment (bizzappdevteam) [inline bad_connector_woocommerce/models/product/importer.py]: let's discuss the approach.

File: bad_connector_woocommerce/models/product/importer.py [hunk 5]
```diff
@@ -199,6 +199,52 @@ def stock_status(self, record):
...
+
+    def _find_existing_downloadable(self, file_id, url):
+        """
+        Find an existing downloadable product record based on file_id and URL.
+        :param file id: The id of the downloadable.
+        :param url: The URL of the downloadable.
+        :return: Existing downloadable product record or None if not found.
+        """
+        return self.env["woo.downloadable.product"].search(
+            [("file_id", "=", file_id), ("url", "=", url)], limit=1
```
Comment (bizzappdevteam) [inline bad_connector_woocommerce/models/product/importer.py]: also search base on "name"

File: bad_connector_woocommerce/models/product/importer.py [hunk 6]
```diff
@@ -199,6 +199,53 @@ def stock_status(self, record):
...
+        :param download_product_info: Information about the downloadable product.
+        :return: downloadable product record.
+        """
+        file_id, name, url = (
+            download_product_info.get("id"),
+            download_product_info.get("name"),
+            download_product_info.get("file"),
+        )
+        existing_downloadable = self._find_existing_downloadable(file_id, name, url)
+        return existing_downloadable or self.env["woo.downloadable.product"].create(
```
Comment (bizzappdev) [inline bad_connector_woocommerce/models/product/importer.py]: why not child import?

File: bad_connector_woocommerce/README.rst [hunk 7]
```diff
@@ -56,6 +56,7 @@
     - Added 'Product Category' field which is located at Connectors> WooCommerce > Advanced Configuration which is use to Set Odoo Product Category for imported WooCommerce Product.
     - Added 'Default Product Type' field which is located at Connectors> WooCommerce > Advanced Configuration which is use to Set Odoo Product Type for imported WooCommerce Product.
     - Added 'WooCommerce Product Image URL' which is located at Product Binding level, designed to store Other Product Images which will store in woo.product.image.url object instead of initial Image.
+    - Added 'WooCommerce Downloadable Product' which is located at Product Binding level, designed to store Downloadable Product files which will store in woo.downloadable.product object.
```
Comment (bizzappdevteam) [inline bad_connector_woocommerce/README.rst]: also update the readme by adding information regarding woo.downloadable.product

File: bad_connector_woocommerce/models/product/importer.py [hunk 8]
```diff
@@ -199,6 +207,43 @@ def stock_status(self, record):
...
+        downloadable_in_record = ()
+        if product_id:
+            product = product_binder.to_internal(product_id)
+            if product:
+                downloadable_in_record = product.woo_downloadable_product_ids.mapped(
+                    "external_id"
+                )
+        removable_product = set(product_ids).symmetric_difference(
+            downloadable_in_record
+        )
```
Comment (bizzappdevteam) [inline bad_connector_woocommerce/models/product/importer.py]: downloadable_in_record = set()
    
    if product_id:
        product = product_binder.to_internal(product_id)
        if product:
            downloadable_in_record = set(
                product.woo_downloadable_product_ids.mapped("external_id")
            )

    removable_product = set(product_ids) ^ downloadable_in_record

File: bad_connector_woocommerce/models/woo_downloadable_product/common.py [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+class WooDownloadableProduct(models.Model):
+    _name = "woo.downloadable.product"
+    _inherit = ["woo.binding"]
+    _description = "WooCommerce Downloadable Product"
+
+    name = fields.Char(string="File Name", required=True)
+    url = fields.Char(string="File URL")
+    woo_product_id = fields.Many2one(
+        "woo.product.product",
+        ondelete="restrict",
```
Comment (bizzappdev) [inline bad_connector_woocommerce/models/woo_downloadable_product/common.py]: cascade

**Comment by bizzappdevteam** [comment]
Closing this PR due to this functionality is there in https://github.com/BizzAppDev-Systems/connector-woocommerce/pull/71

**Pull request closed without merging**